Konami Code (Workshop Bounty) (#1346)
hackclub/hackclub: workshops/konami_code/README.md

Here, the `secretCode` array contains the keys which will be the Konami Code. For this workshop's sake, our code will be "hack". You can change it to anything you like. Here are some examples:

```js
["ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight"]
["ArrowUp", "a", "ArrowLeft", "z", "ArrowDown"]
["Shift", "Control", "Enter", "b"]
["a", "a", "b", "b", "c", "d", "e", "ArrowLeft"]
```

Next, we will make use of the confetti.js library which we have imported in our project. We will first declare some settings for the confetti. After we declare them, we will create a new `ConfettiGenerator` object which will help us implement the confetti on the site.

```js
const confettiSettings = { target: 'my-canvas', size: 1.4, max: 150 };
const confetti = new ConfettiGenerator(confettiSettings);
```
**Explanation:** The settings for the confetti is an object. It is made with reference to [this](https://github.com/Agezao/confetti-js/) site. There are many more options which you can pass to the settings as you wish.

The `target` attribute is used to specify the canvas with the id `my-canvas` on which the confetti will appear.

The `size` attribute specifies the size for the confetti particles.

> The `max` attribute specifies the The number of props(confetti) to be rendered.

Reviewer: ```suggestion
The `max` attribute specifies The number of props(confetti) to be rendered.
```

---
name: 'Konami Code'
description: 'Implement a konami code on a website'
author: '@faisalsayed10'
---

# Konami Code

> You may have heard about movie Easter Eggs or video game Easter Eggs before, but have you ever heard about website Easter Eggs? The internet is filled with copious amounts of weird and wonderful hidden features! I bet you have played the popular game GTA, right? GTA too had konami codes but they were called 'cheat codes'. 

Reviewer: ```suggestion

You may have heard about the movie Easter Eggs or video game Easter Eggs before, but have you ever heard about the website Easter Eggs? The internet is filled with copious amounts of weird and wonderful hidden features! I bet you have played the popular game GTA, right? GTA too had Konami codes but they were called 'cheat codes'. 
```

Reviewer 2: in this context `the` is not needed, but @faisalsayed10 you might want to make Easter eggs lower case

Author: Where have u found this file?
I think this is my very old file, and this isn't what is in the PR

Author: @bezlin6mechminerz you have reviewed my very old commit, this is not which is the real PR

If you don't know what they are, check out these cool easter eggs/konami codes on the internet:

1. Try searching 'do a barrel roll' on google.

2. Try inspecting [Max's Website](maxwofford.com) 🤨

3. Konami Code On [BuzzFeed](https://www.buzzfeed.com/) (Just press up, up, down, down, left, right, left, right, B, A and enter on the homepage).

4. In the Firefox browser bar, type "about:mozilla" or "about:robots" to see the secrets of Mozilla.

> Well, by now, you must have known what a konami code is. How about we build our own konami code?

Reviewer: ```suggestion
If you don't know what they are, check out these cool easter eggs/Konami codes on the internet:

1. Try searching 'do a barrel roll' on google.

2. Try inspecting [Max's Website](maxwofford.com) 🤨

3. Konami Code On [BuzzFeed](https://www.buzzfeed.com/) (Just press up, up, down, down, left, right, left, right, B, A, and enter on the homepage).

4. In the Firefox browser bar, type "about:mozilla" or "about:robots" to see the secrets of Mozilla.

Well, by now, you must have known what a Konami code is. How about we build our own Konami code?
```

It will look something like this after someone types the konami code:

[![Image](https://cloud-5apj3kt1e.vercel.app/image.png)](https://konami-code.faisalsayed1.repl.co/)

Here's the [live demo](https://konami-code.faisalsayed1.repl.co/)  
Here's the [source code](https://repl.it/@FaisalSayed1/Konami-Code)

Live demo: 
![Image](https://cloud-95l2730mf.vercel.app/konamigif.gif)

> 

Reviewer: ```suggestion

It will look something like this after someone types the Konami code:

[![Image](https://cloud-5apj3kt1e.vercel.app/image.png)](https://konami-code.faisalsayed1.repl.co/)

Here's the [live demo](https://konami-code.faisalsayed1.repl.co/)  
Here's the [source code](https://repl.it/@FaisalSayed1/Konami-Code)

Live demo: 
![Image](https://cloud-95l2730mf.vercel.app/konamigif.gif)

```

Author: What are the changes here? there is nothing different? why then? a review?

## Part 1: Prerequisites

You should have a beginner understanding of:

- HTML
- CSS
- JavaScript

Also for this workshop, we will need a library called 'confetti.js'. 

Download it by going [here](https://cloud-72a7nfmm7.vercel.app/index.min.js). Right click and select 'save as'.

> 

Reviewer: ```suggestion

You should have a beginner understanding of:

- HTML
- CSS
- JavaScript

Also for this workshop, we will need a library called 'confetti.js'. 

Download it by going [here](https://cloud-72a7nfmm7.vercel.app/index.min.js). Right-click and select 'save as'.

```

Author: same here too? what's changed here?

## Part 2: Setup

### Setting up your code environment on Repl.it

[Repl.it](https://repl.it) is an online code editor. You don't have to use Repl.it but I suggest you do as it sets everything up for you and you don't require any installations.

To get started, go to [https://repl.it/languages/html](https://repl.it/languages/html). Your coding environment will spin up in just a few seconds!

You should see something like the following:

![Image](https://cloud-p7qnbqzo6.vercel.app/image.png)

Import the confetti.js (`index.min.js`) file which you downloaded on the left side of the screen.

![Image](https://cloud-2d72c0n78.vercel.app/indexminjs.gif)

## Part 3: Building the project

### 1) HTML

Let's start writing the HTML code required in the `index.html` file. First, we'll import the required JavaScript files.

Inside the `<head>` tag, we'll import the [`anime.js`](https://animejs.com/) library. Copy the code given below.

```html
<script src="https://cutt.ly/anime_js"></script>
```

Also link the `index.min.js` file to our HTML inside the `<head>` tag.

```html
<script src="index.min.js"></script>
```

The code so far:

> 

Reviewer: ```suggestion

Let's start writing the HTML code required in the `index.html` file. First, we'll import the required JavaScript files.

Inside the `<head>` tag, we'll import the [`anime.js`](https://animejs.com/) library. Copy the code given below.

```html
<script src="https://cutt.ly/anime_js"></script>
```

Also, link the `index.min.js` file to our HTML inside the `<head>` tag.

```html
<script src="index.min.js"></script>
```

The code so far:

```

Author: To be honest, you have confused me so much

Reviewer: oh sorry
i just fixed the typos

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width">
    <title>repl.it</title>
    <link href="style.css" rel="stylesheet"
    type="text/css" />
    <script src="https://cutt.ly/anime_js"></script>
    <script src="index.min.js"></script>
  </head>
  <body>
    <script src="script.js"></script>
  </body>
</html>
```

Next, we'll add some basic HTML code inside of our `<body>` tag.

```html
<canvas id="my-canvas"></canvas>
<h1 class='text'>Password Please</h1>
<div class="secret-div">
  <h1>Hello comrade!</h1>
</div>
```

**Note:** Make sure the `<script>` tag, which is in your body, is just above the closing `</body>` tag and whatever code you write in the `<body>` should be above that `<script>` tag or else it will throw you errors.

**Explanation:** We are adding a canvas element which will be later used for the confetti. Then there is also a `secret-div`, it will be initially hidden but when someone types the konami code, we'll make it visible.

> 

Reviewer: ```suggestion
**Note:** Make sure the `<script>` tag, which is in your body, is just above the closing `</body>` tag and whatever code you write in the `<body>` should be above that `<script>` tag or else it will throw you errors.

**Explanation:** We are adding a canvas element which will be later used for the confetti. Then there is also a `secret-div`, it will be initially hidden but when someone types the Konami code, we'll make it visible.

```

Author: same here too. why?

```js
["KeyUp", "KeyDown", "KeyLeft", "KeyRight"]
["KeyUp", "a", "KeyLeft", "z", "KeyDown"]
["Shift", "Control", "Enter", "b"]
["a", "a", "b", "b", "c", "d", "e", "KeyLeft"]
```

Next, we will make use of the confetti.js library which we have imported in our project. We will first declare some settings for the confetti. After we declare them, we will create a new `ConfettiGenerator` object which will help us implement the confetti on the site.

```js
const confettiSettings = { target: 'my-canvas', size: 1.4, max: 150 };
const confetti = new ConfettiGenerator(confettiSettings);
```
**Explanation:** The settings for the confetti is an object. It is made with reference to [this](https://github.com/Agezao/confetti-js/) site. There are many more options which you can pass to the settings as you wish.

The `target` attribute is used to specify the canvas with the id `my-canvas` on which the confetti will appear.

The `size` attribute specifies the size for the confetti particles.

The `max` attribute specifies the The number of props(confetti) to be rendered.

The `ConfettiGenerator` contains a `render()` method and when it is called, we will be able to see confetti on our page.

> For more information on these attributes, [check here](https://github.com/Agezao/confetti-js/).

Reviewer: ```suggestion
**Explanation:** The setting for the confetti is an object. It is made with reference to [this](https://github.com/Agezao/confetti-js/) site. There are many more options than you can pass to the settings as you wish.

The `target` attribute is used to specify the canvas with the id `my-canvas` on which the confetti will appear.

The `size` attribute specifies the size for the confetti particles.

The `max` attribute specifies The number of props(confetti) to be rendered.

The `ConfettiGenerator` contains a `render()` method and when it is called, we will be able to see confetti on our page.

For more information on these attributes, [check here](https://github.com/Agezao/confetti-js/).
```

Your code so far:
```js
const text = document.querySelector(".text");
const secretDiv = document.querySelector('.secret-div');

const keys = [];
const secretCode = ["h", "a", "c", "k"];

const confettiSettings = { target: 'my-canvas', size: 1.4, max: 150 };
const confetti = new ConfettiGenerator(confettiSettings);
```

> Next we will implement a function to check which keys are pressed by the user and when the keys match our Konami Code, it will get triggered.
> 
> The function will take an 'event' attribute which will give us the key pressed by the user.
> 

Reviewer: ```suggestion
Next, we will implement a function to check which keys are pressed by the user, and when the keys match our Konami Code, it will get triggered.

The function will take an 'event' attribute which will give us the key pressed by the user.

```

```js
function keysDetector(e) {
  console.log(e.key);
}
```

Outside the function, we will add an event listener on the window which will listen for all the `keyup` events and then trigger a particular function. It is written as follows:

```js
window.addEventListener('keyup', keysDetector);
```

**Explanation:** `addEventListener()` takes the first parameter as the event of which we need to keep track of. In our case it is the `keyup` event. The second parameter is a function which will get called whenever the event specified to it is performed. So whenever the `keyup` event is performed, the `keysDetector` function gets called.

> Learn more about [`addEventListener()`](https://www.w3schools.com/js/js_htmldom_eventlistener.asp).
> 
> 'RUN' this code, click on the preview window once so that it gets focused and then try typing any keys. Keep a watch on the console!
> 
> Voila! The keys pressed by the user is getting logged! We will make use of this keys to check whether the keys matches the Konami Code.

Reviewer: ```suggestion
Learn more about [`addEventListener()`](https://www.w3schools.com/js/js_htmldom_eventlistener.asp).

'RUN' this code, click on the preview window once so that it gets focused and then try typing any keys. Keep a watch on the console!

Voila! The keys pressed by the user is getting logged! We will make use of these keys to check whether the keys match the Konami Code.
```

Try console logging the `keys` array now (inside the function) and you'll get a better understanding of the above code.

And now, the time has come! We will compare both the arrays and whenever `keys === secretCode` we will do some magic!

But wait a minute...

![Image](https://media.giphy.com/media/RkKMFRHh71RvhCYUx6/giphy.gif)

Unfortunately, we can't simply write `keys === secretCode` and get away with it :( The reason for this is as arrays are objects in JavaScript, the triple equals operator `(===)` only returns true if the arrays are the same reference.

Example:
```js
const a = [1, 2, 3];
const b = [1, 2, 3];

a === a; // true
a === b; // false
```

So, we'll need to go another way which is known as...  
**`JSON.stringify()`**!

The JSON.stringify() method will convert a JavaScript object or value to a JSON string. So, we will first convert both the arrays to a string and then, we are free to compare it!

Here's how to do it:
```js
if (JSON.stringify(keys) === JSON.stringify(secretCode)) {
  //do something or lets just console.log for now.
  console.log('TADAAA!!');
}
```

> **Note:** This will also be included in the function. Except the event listener, nothing is outside the function.
> 
> RUN the code and try typing the secret code which you have given (it is "hack" in my case). You'll surely see a TADAAA!! in the console whenever the arrays match each other.
> 
> Okay, so now we are sure that our code works perfectly! So, instead of console logging we will make some pretty changes to our website.
> 
> First of all, we'll make our `secret-div` visible. Then we'll change the `text` from 'Password Please' to 'Identity Verified'. And not to forget, we'll also render confetti on our page.
> 
> **Challenge:** Can you do this all on your own?  
> **Note:** Do it inside the `if` statement and also remove the `console.log()`.  
> **Hints:**  

Reviewer: ```suggestion
**Note:** This will also be included in the function. Except for the event listener, nothing is outside the function.

RUN the code and try typing the secret code which you have given (it is "hack" in my case). You'll surely see a TADAAA!! in the console whenever the arrays match each other.

Okay, so now we are sure that our code works perfectly! So, instead of console logging, we will make some pretty changes to our website.

First of all, we'll make our `secret-div` visible. Then we'll change the `text` from 'Password Please' to 'Identity Verified'. And not to forget, we'll also render confetti on our page.

**Challenge:** Can you do this all on your own?  
**Note:** Do it inside the `if` statement and also remove the `console.log()`.  
**Hints:**  
```
